cpp/src/arrow/compute/kernels/codegen_internal.cc

```diff
@@ -251,6 +251,8 @@ TypeHolder CommonTemporal(const TypeHolder* begin, size_t count) {
   bool saw_date32 = false;
   bool saw_date64 = false;
   bool saw_duration = false;
+  bool saw_time32 = false;
+  bool saw_time64 = false;
```

Comment:
This could be simplified since time32 only supports seconds and milliseconds whereas time64 only supports microseconds and nanoseconds
```suggestion
  bool saw_time_since_midnight = false;
```

IE if you have seen a time type and finest_unit is `SECOND` or `MILLI` then you'll be outputting time32, otherwise time64